R/S4.R

```diff
@@ -1,3 +1,19 @@
+#' Register an R7 class with S4
+#'
+#' If you want to use [method<-] to register an method for an S4 generic with
+#' an R7 class, you need to call `S4_register()` once.
+#'
+#' @param class An R7 class created with [new_class()].
+#' @param env Expert use only. Environment where S4 class will be registered.
+#' @export
+S4_register <- function(class, env = parent.frame()) {
+  if (!is_class(class)) {
+    msg <- sprintf("`class` must be an R7 class, not a %s", obj_desc(class))
+  }
+
+  methods::setOldClass(class_dispatch(class), where = topenv(env))
```

Comment:
One advantage would be that `validObject()` could check the types of the "slots". Passing an S4 class definition also lets us pass our own validity function, as well as a prototype, which S4 would use for the default value of a slot. This may also come in handy if we come to support extending R7 with S4.

Reply:
This seems to be non-trivial so I'm going to merge this PR and start work on full class definition in another PR.